File: lib/cpp/tests/ProxyTest.cpp
#include <gtest/gtest.h>
#include <zddm/Proxy.h>
#include <zddm/adapters/InMemoryStorageAdapter.h>
#include <zddm/hashers/SimpleHasher.h>

#include <optional>
#include <unordered_map>

namespace {

// Always open traffic-gate to make the tests easier
std::unique_ptr<zddm::TrafficGate<std::string>> createAlwaysOnGate() {
  auto hasher = std::make_unique<zddm::hashers::SimpleHasher<std::string>>();
  return std::make_unique<zddm::TrafficGate<std::string>>(
      1.0 /* rate = 1.0 means always pass */, std::move(hasher));
}

std::unique_ptr<
    zddm::adapters::InMemoryStorageAdapter<std::string, std::string>>
createTestStorageAdapter() {
  return std::make_unique<
      zddm::adapters::InMemoryStorageAdapter<std::string, std::string>>();
}

}  // namespace

namespace zddm::tests {

TEST(ProxyTests, TestIsEnabled) {
  zddm::Proxy<std::string, std::string> proxy(
      createAlwaysOnGate(),
      /* old */ createTestStorageAdapter(),
      /* new */ createTestStorageAdapter());

  // Enabled by default
  ASSERT_TRUE(proxy.isEnabled());
  // On disable, should be disabled
  ASSERT_FALSE(proxy.disable().isEnabled());
  // On enable, should be disabled
  ASSERT_TRUE(proxy.enable().isEnabled());
}

TEST(ProxyTests, TestShouldAlwaysReadOldOnDisabled) {
  auto old_adapter = createTestStorageAdapter();
  auto new_adapter = createTestStorageAdapter();

  // Make sure the same key exists for both storage solutions, with
  // different data [for testing purposes].
  old_adapter->write("foo", "old_bar");
  new_adapter->write("foo", "bar");

  zddm::Proxy<std::string, std::string> proxy(
      createAlwaysOnGate(), std::move(old_adapter), std::move(new_adapter));

  proxy.disable();
  ASSERT_EQ(proxy.read("foo"), "old_bar");
}

TEST(ProxyTests, TestShouldAlwaysWriteOldOnDisabled) {
  auto old_adapter = createTestStorageAdapter();
  auto new_adapter = createTestStorageAdapter();

  // Make sure the same key exists for both storage solutions, with
  // different data [for testing purposes].
  old_adapter->write("foo", "bar");
  new_adapter->write("foo", "bar");

  zddm::Proxy<std::string, std::string> proxy(
      createAlwaysOnGate(), std::move(old_adapter), std::move(new_adapter));

  proxy.disable();
  proxy.write("foo", "old");
  ASSERT_EQ(proxy.read("foo"), "old");
}

TEST(ProxyTests, TestShouldReadWithPriorityForNew) {
  auto old_adapter = createTestStorageAdapter();
  auto new_adapter = createTestStorageAdapter();

  // Make sure the same key exists for both storage solutions, with
  // different data [for testing purposes].
  old_adapter->write("foo", "old_bar");
  new_adapter->write("foo", "bar");

  zddm::Proxy<std::string, std::string> proxy(
      createAlwaysOnGate(), std::move(old_adapter), std::move(new_adapter));

  // As we access "foo" we should hit the new storage first by default
  // and get the value stored in the new adapter rather than in the
  // old.
  EXPECT_EQ(proxy.read("foo"), "bar");
}

TEST(ProxyTests, TestShouldReadOldOnMiss) {
  auto old_adapter = createTestStorageAdapter();
  auto new_adapter = createTestStorageAdapter();

  // Make sure the same key exists for both storage solutions, with
  // different data [for testing purposes].
  old_adapter->write("foo", "old_foo");

  zddm::Proxy<std::string, std::string> proxy(
      createAlwaysOnGate(), std::move(old_adapter), std::move(new_adapter));

  // As we access "foo" we should hit the new storage first by default
  // and get the value stored in the new adapter rather than in the
  // old.
  EXPECT_EQ(proxy.read("foo"), "old_foo");
}

TEST(ProxyTests, TestShouldReadOldOnAndCopyOnMiss) {
  auto old_adapter = createTestStorageAdapter();
  auto new_adapter = createTestStorageAdapter();

  // Make sure we only have one value that produces a hit but in the
  // old storage
  old_adapter->write("foo", "bar");

  // Grab a reference so we can check if it exists in storage
  // after handing over ownership of `new_adapter`.
  auto const& new_storage = new_adapter->getWeakPointerToContainer();

  zddm::Proxy<std::string, std::string> proxy(
      createAlwaysOnGate(), std::move(old_adapter), std::move(new_adapter));

  // First read should hit the old storage successfully
  EXPECT_EQ(proxy.read("foo"), "bar");
  // New read should hit the new storage successfully
  EXPECT_EQ(proxy.read("foo"), "bar");

  // Make sure we have the value in new storage, as expected
  EXPECT_TRUE(new_storage->find("foo") != new_storage->end());
  // Adjust the value to make them diverge, doubling down on that
  // we prioritise the new storage
  new_storage->insert_or_assign("foo", "new_bar");
  // New read should hit the new storage successfully
  EXPECT_EQ(proxy.read("foo"), "new_bar");
}

TEST(ProxyTests, TestShouldDoubleWrite) {
  auto old_adapter = createTestStorageAdapter();
  auto new_adapter = createTestStorageAdapter();

  // Make sure we only have one value that produces a hit but in the
  // old storage
  old_adapter->write("foo", "bar");

  // Grab references so we can check if it exists in storage
  // after handing over ownership of `new_adapter`.
  auto const& old_storage = new_adapter->getWeakPointerToContainer();
  auto const& new_storage = new_adapter->getWeakPointerToContainer();

  zddm::Proxy<std::string, std::string> proxy(
      createAlwaysOnGate(), std::move(old_adapter), std::move(new_adapter));

  // Updating "foo" should write the update to both storage solutions
  proxy.write("foo", "new_bar");

  // Both values should be present in both storages
  EXPECT_EQ(old_storage->at("foo"), "new_bar");
  EXPECT_EQ(new_storage->at("foo"), "new_bar");
}

}  // namespace zddm::tests

File: lib/go/zddm/proxy_test.go
package zddm

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// This is a hack as we're not really supposed
// to store primitives, but pointers to complex objects
func newString(s string) *string {
	return &s
}

func createAlwaysOnGate() *TrafficGate[string] {
	return NewTrafficGate[string](
		1.0, /* rate = 1.0 means always pass */
		&simpleHasher{},
	)
}

func createTestStorageAdapter() *InMemoryStorageAdapter[string, string] {
	return &InMemoryStorageAdapter[string, string]{
		container_: make(map[string]*string),
	}
}

func TestProxyIsEnabled(t *testing.T) {
	assert := assert.New(t)
	proxy := NewProxy[string, string](
		createAlwaysOnGate(),
		/* old */ createTestStorageAdapter(),
		/* new */ createTestStorageAdapter(),
	)

	assert.True(proxy.IsEnabled(), "Proxy should be enabled by default")
	proxy.Disable()
	assert.False(proxy.IsEnabled(), "On disable, should be disabled")
	proxy.Enable()
	assert.True(proxy.IsEnabled(), "On enable, should be enabled")
}

func TestProxyShouldAlwaysReadOldOnDisabled(t *testing.T) {
	old_adapter := createTestStorageAdapter()
	new_adapter := createTestStorageAdapter()

	// Make sure the same key exists for both storage solutions, with
	// different data [for testing purposes].
	old_adapter.Write("foo", newString("old_bar"))
	new_adapter.Write("foo", newString("bar"))

	proxy := NewProxy[string, string](
		createAlwaysOnGate(),
		old_adapter,
		new_adapter,
	)

	proxy.Disable()
	assert.Equal(t, *proxy.Read("foo"), "old_bar", "We should have read from old storage")
}

func TestProxyShouldAlwaysWriteOldOnDisabled(t *testing.T) {
	old_adapter := createTestStorageAdapter()
	new_adapter := createTestStorageAdapter()

	// Make sure the same key exists for both storage solutions, with
	// different data [for testing purposes].
	old_adapter.Write("foo", newString("old_bar"))
	new_adapter.Write("foo", newString("bar"))

	proxy := NewProxy[string, string](
		createAlwaysOnGate(),
		old_adapter,
		new_adapter,
	)

	proxy.Disable()
	proxy.Write("foo", newString("old"))
	assert.Equal(t, *proxy.Read("foo"), "old", "We should have updated the value correctly")

	// Make sure we did not write to new
	assert.Equal(t, *new_adapter.Read("foo"), "bar", "We should have retained the value in the new storage")
}

func TestProxyShouldReadWithPriorityForNew(t *testing.T) {
	old_adapter := createTestStorageAdapter()
	new_adapter := createTestStorageAdapter()

	// Make sure the same key exists for both storage solutions, with
	// different data [for testing purposes].
	old_adapter.Write("foo", newString("old_bar"))
	new_adapter.Write("foo", newString("bar"))

	proxy := NewProxy[string, string](
		createAlwaysOnGate(),
		old_adapter,
		new_adapter,
	)

	assert.Equal(t, *proxy.Read("foo"), "bar", "We should read from the new storage if the key exists")
}

func TestProxyShouldReadOldOnMiss(t *testing.T) {
	old_adapter := createTestStorageAdapter()
	new_adapter := createTestStorageAdapter()

	// We should only have one value in the old storage
	old_adapter.Write("foo", newString("old_bar"))

	proxy := NewProxy[string, string](
		createAlwaysOnGate(),
		old_adapter,
		new_adapter,
	)

	assert.Equal(t, *proxy.Read("foo"), "old_bar", "We should have read from old storage")
}

func TestProxyShouldReadOldAndCopyOnMiss(t *testing.T) {
	assert := assert.New(t)
	old_adapter := createTestStorageAdapter()
	new_adapter := createTestStorageAdapter()

	// We should only have one value in the old storage
	old_adapter.Write("foo", newString("bar"))

	proxy := NewProxy[string, string](
		createAlwaysOnGate(),
		old_adapter,
		new_adapter,
	)

	// First read should hit old storage successfully
	assert.Equal(*proxy.Read("foo"), "bar", "We should have read from old storage")
	// Second read should hit new storage successfully
	assert.Equal(*proxy.Read("foo"), "bar", "We should have read from new storage [again]")

	// Make sure the value has been pushed into the new storage
	newData := new_adapter.Read("foo")
	assert.NotNil(newData, "We should have the value in new storage")
	assert.Equal(*newData, "bar", "We should have written a copy to new storage")
	// Adjust the value to make them diverge, doubling down on that
	// we prioritise the new storage
	new_adapter.Write("foo", newString("new_bar"))
	assert.Equal(*proxy.Read("foo"), "new_bar", "We should have read from new storage")
}

func TestProxyShouldDoubleWrite(t *testing.T) {
	assert := assert.New(t)
	old_adapter := createTestStorageAdapter()
	new_adapter := createTestStorageAdapter()

	// We should only have one value in the old storage
	old_adapter.Write("foo", newString("bar"))

	proxy := NewProxy[string, string](
		createAlwaysOnGate(),
		old_adapter,
		new_adapter,
	)

	// Updating "foo" should write the update to both storage solutions
	proxy.Write("foo", newString("new_bar"))

	// Both values should be updated
	assert.Equal(*old_adapter.Read("foo"), "new_bar", "Should have updated old storage")
	assert.Equal(*new_adapter.Read("foo"), "new_bar", "Should have updated new storage")
}
